src/exchanges/populate.ts

```diff
@@ -0,0 +1,331 @@
+      forward
+    );
+  };
+};
+
+type UserFragmentMap<T extends string = string> = Record<
+  T,
+  FragmentDefinitionNode
+>;
+
+type TypeFragmentMap<T extends string = string> = Record<T, TypeFragment[]>;
+
+interface TypeFragment {
+  /** Operation key where selection set is being used. */
+  key: number;
+  /** Selection set. */
+  fragment: FragmentDefinitionNode;
+  /** Type of selection. */
+  type: string;
+}
+
+interface MakeFragmentsFromQueryArg {
+  schema: GraphQLSchema;
+  query: DocumentNode;
+}
+
+/** Gets typed selection sets and fragments from query */
+export const extractSelectionsFromQuery = ({
+  schema,
+  query,
+}: MakeFragmentsFromQueryArg) => {
+  let selections: { selections: SelectionSetNode; type: string }[] = [];
+  let fragments: FragmentDefinitionNode[] = [];
+  const typeInfo = new TypeInfo(schema);
+
+  visit(
+    query,
+    visitWithTypeInfo(typeInfo, {
+      Field: node => {
+        if (!node.selectionSet) {
+          return undefined;
+        }
+
+        const type = getType(typeInfo);
+
+        if (!type) {
+          return undefined;
+        }
+
+        selections = [...selections, { selections: node.selectionSet, type }];
+      },
+      FragmentDefinition: node => {
+        fragments = [...fragments, node];
+      },
+    })
+  );
+
+  return { selections, fragments };
+};
+
+interface AddFragmentsToQuery {
+  schema: GraphQLSchema;
+  query: DocumentNode;
+  typeFragments: Record<string, Omit<TypeFragment, 'key'>[]>;
+  userFragments: UserFragmentMap;
+}
+
+/** Replaces populate decorator with fragment spreads + fragments. */
+export const addFragmentsToQuery = ({
+  schema,
+  query,
+  typeFragments,
+  userFragments,
+}: AddFragmentsToQuery) => {
+  const typeInfo = new TypeInfo(schema);
+  const requiredUserFragments = new Set<FragmentDefinitionNode>();
+  let additionalFragments: Record<string, FragmentDefinitionNode> = {};
+
+  return visit(
+    query,
+    visitWithTypeInfo(typeInfo, {
+      Field: {
+        enter: node => {
+          if (
+            !node.directives ||
+            !node.directives.find(d => d.name.value === 'populate')
+          ) {
+            return;
+          }
+
+          const directives = node.directives.filter(
+            d => d.name.value !== 'populate'
+          );
+
+          const types = getTypes(schema, typeInfo);
+          const newSelections = types.reduce((p, t) => {
+            const typeFrags = typeFragments[t.name];
+            if (!typeFrags) {
+              return p;
+            }
+
+            return [
+              ...p,
+              ...typeFragments[t.name].map(({ fragment }) => {
+                // Add used fragment for insertion at Document node
+                getUsedFragments(fragment).forEach(f =>
+                  requiredUserFragments.add(userFragments[f])
+                );
+
+                // Add fragment for insertion at Document node
+                additionalFragments = {
```

Comment:
There's a lot of local spreading/copying in this file (inside reduces, here as well, etc). Generally I don't mind that at all, but it all adds up in terms of extra-allocations which we can avoid to make this run nice and fast :)

Reply:
Sounds like Map is what we want to be using